### PR TITLE
[Misc] centralize all usage of environment variables

### PR DESCRIPTION
**Pull request by youkaichao**
An implementation of RFC https://github.com/vllm-project/vllm/issues/4407 .

Exceptions are:

- Test code related environment variables, leave as they are
- Modify/write to environment variables,  leave as they are
- environment variables used in `setup.py`, when we cannot import `vllm/envs.py` . leave as they are

### CODE REVIEW

File: vllm/distributed/utils.py
Comment (sfc-gh-zhwang) [inline vllm/distributed/utils.py:111]: can we use a different env var for these cache like files vs library-like files(libnccl.so, etc)?

Comment (youkaichao) [reply vllm/distributed/utils.py:111]: Any good reason? We already have so many env vars, I want to reduce the amount if possible 🙏 

If you want to separate them, you can use `VLLM_NCCL_SO_PATH` to point to the library file anyway.

Comment (sfc-gh-zhwang) [reply vllm/distributed/utils.py:111]: basically the cache file is written **after** vllm engine is started **by** the engine itself, the `.so` file are built by docker and considered **immutable**.

Companies usually disable root file system's write access, so they will have to mount a directory to `VLLM_CONFIG_ROOT`, but doing this will nuke all the `.so` files in that directory.

Comment (youkaichao) [reply vllm/distributed/utils.py:111]: My suggestion:
1. mount a directory, and set `VLLM_CONFIG_ROOT` to that directory
2. copy the original `/root/.config` to that directory, change permission for the user you use.
Then you are all set.

Comment (sfc-gh-zhwang) [reply vllm/distributed/utils.py:111]: your suggestion or other workaround can solve this, but semantically i don't think this `cache` files are part of the config.

Comment (youkaichao) [reply vllm/distributed/utils.py:111]: in the long run, we hope nccl team can solve the problem, so we don't need to manipulate library this way.

File: vllm/envs.py
Comment (rkooo567) [inline vllm/envs.py:155]: just not familiar with this python magic here. Would this trigger lambda only once or it is called whenever we access the attribute?

**Review by rkooo567 — APPROVED** [review]
LGTM if tests pass

File: vllm/envs.py
Comment (youkaichao) [reply vllm/envs.py:155]: it is called whenever we access the attribute. this is strictly equivalent to how we use `os.getenv`, because we want it to read the latest env var value.

**Review by sfc-gh-zhwang — APPROVED** [review]

File: vllm/envs.py
Comment (rkooo567) [reply vllm/envs.py:155]: I see. yeah this is also good for testing.

**Review by simon-mo — APPROVED** [review]

**Comment by simon-mo** [comment]
Oh can you do a follow up adding this to docs, you can use `literalinclude` directive (search this term in docs).

**Pull request merged**
